c++ training/LinkedLists/LinkedLists/DoublyLinkedList.h

```diff
@@ -0,0 +1,323 @@
+	{
+		return GetNodeAt(Index).Data;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	T& GetHead()
+	{
+		return Head->Data;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	const T& GetHead() const
+	{
+		return Head->Data;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	T& GetTail()
+	{
+		return Tail->Data;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	const T& GetTail() const
+	{
+		return Tail->Data;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	bool IsEmpty() const
+	{
+		return Size == 0;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	int GetSize() const
+	{
+		return Size;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	void AddHead(const T& Data)
+	{
+		if (IsEmpty())
+		{
+			Head = new FNode(Data);
+			Tail = Head;
+
+			Size = 1;
+			return;
+		}
+
+		Head->Previous = new FNode(Data, Head);
+		Head = Head->Previous;
+
+		Size++;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	void AddTail(const T& Data)
+	{
+		if (IsEmpty())
+		{
+			AddHead(Data);
+			return;
+		}
+
+		Tail->Next = new FNode(Data, nullptr, Tail);
+		Tail = Tail->Next;
+
+		Size++;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	void Insert(const T& Data, const int Index)
+	{
+		if (Index < 0 || Index > Size)
+		{
+			return;
+		}
+
+		if (Index == 0)
+		{
+			AddHead(Data);
+			return;
+		}
+		else
+		{
+			FNode& PreviousNode = GetNodeAt(Index - 1);
+			FNode* CurrentPtr = PreviousNode.Next;
+			PreviousNode.Next = new FNode(Data, CurrentPtr, &PreviousNode);
+
+			if (Index == Size)
+			{
+				Tail = Tail->Next;
+			}
+			else
+			{
+				CurrentPtr->Previous = PreviousNode.Next;
+			}
```

Comment:
here you can use GetNodeAt(Index); there is no more need of the Index - 1, it will change some logic but is more clear

Reply:
The thing is, I'm trying to also group up the case in which Index can actually be the same as Size, so I can't in fact call GetNodeAt on such index, which is the main reason i act on the previous one for this is guaranteed to be an actual node